arangod/RestHandler/RestCursorHandler.cpp

```diff
@@ -659,6 +676,68 @@ RestStatus RestCursorHandler::createQueryCursor() {
   return registerQueryOrCursor(body);
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// @brief shows the batch given by <batch-id> if it's the last cached batch
+/// response on a retry, and does't advance cursor
+////////////////////////////////////////////////////////////////////////////////
+
+RestStatus RestCursorHandler::showLatestBatch() {
+  std::vector<std::string> const& suffixes = _request->suffixes();
+
+  if (suffixes.size() != 2) {
+    generateError(rest::ResponseCode::BAD, TRI_ERROR_HTTP_BAD_PARAMETER,
+                  "expecting POST /_api/cursor/<cursor-id>/<batch-id>");
+    return RestStatus::DONE;
+  }
+
+  std::string const& id = suffixes[0];
+
+  auto cursors = _vocbase.cursorRepository();
+  TRI_ASSERT(cursors != nullptr);
+
+  auto cursorId = static_cast<arangodb::CursorId>(
+      arangodb::basics::StringUtils::uint64(id));
+  bool busy;
+  _cursor = cursors->find(cursorId, busy);
+
+  if (_cursor == nullptr) {
+    if (busy) {
+      generateError(GeneralResponse::responseCode(TRI_ERROR_CURSOR_BUSY),
+                    TRI_ERROR_CURSOR_BUSY);
+    } else {
+      generateError(GeneralResponse::responseCode(TRI_ERROR_CURSOR_NOT_FOUND),
+                    TRI_ERROR_CURSOR_NOT_FOUND);
+    }
+    return RestStatus::DONE;
+  }
+  if (!_cursor->isRetriable()) {
+    generateError(rest::ResponseCode::BAD, TRI_ERROR_HTTP_BAD_PARAMETER,
+                  "expecting retriable option to be true");
+    return RestStatus::DONE;
+  }
+
+  _cursor->setWakeupHandler(withLogContext(
+      [self = shared_from_this()]() { return self->wakeupHandler(); }));
+
+  std::shared_ptr<transaction::Context> ctx = _cursor->context();
+
+  VPackBufferUInt8 buffer;
+
+  std::string const& batchId = suffixes[1];
+
+  auto const r = _cursor->getLastBatchResult(batchId, buffer);
```

Comment:
Why yet another Buffer here? Can't we return the shared_ptr to the existing Buffer and simply use it?

Reply:
You can't because, in the following lines, you might have to use `std::move` on the data pointed to by the shared_ptr, and then, you'll have to use it again and doesn't contain the data anymore:
```
 auto const r = _cursor->getLastBatchResult(batchId, buffer);

  if (r.ok()) {
    _response->setContentType(rest::ContentType::JSON);
    generateResult(rest::ResponseCode::OK, std::move(buffer), std::move(ctx));
  }
```
`generateResult()` expects `Payload &&`, rvalue